p2p/adapters/inproc.go

```diff
@@ -112,69 +180,99 @@ func (self *SimNode) setPeer(id *NodeId, rw *p2p.MsgPipeRW) *Peer {
 	return p
 }
 
-func (self *SimNode) Disconnect(rid []byte) error {
+func (self *SimNode) RemovePeer(node *discover.Node) {
 	self.lock.Lock()
 	defer self.lock.Unlock()
-	id := NewNodeId(rid)
+	id := &NodeId{node.ID}
 	peer := self.getPeer(id)
 	if peer == nil || peer.MsgPipeRW == nil {
-		return fmt.Errorf("already disconnected")
+		return
 	}
 	peer.MsgPipeRW.Close()
 	peer.MsgPipeRW = nil
 	// na := self.network.GetNodeAdapter(id)
 	// peer = na.(*SimNode).GetPeer(self.Id)
 	// peer.RW = nil
 	log.Trace(fmt.Sprintf("dropped peer %v", id))
-
-	return nil
 }
 
-func (self *SimNode) Connect(rid []byte) error {
+func (self *SimNode) AddPeer(node *discover.Node) {
 	self.lock.Lock()
 	defer self.lock.Unlock()
-	id := NewNodeId(rid)
+	id := &NodeId{node.ID}
 	na := self.network.GetNodeAdapter(id)
 	if na == nil {
-		return fmt.Errorf("node adapter for %v is missing", id)
+		panic(fmt.Sprintf("node adapter for %v is missing", id))
 	}
 	rw, rrw := p2p.MsgPipe()
 	// // run protocol on remote node with self as peer
 	peer := self.getPeer(id)
 	if peer != nil && peer.MsgPipeRW != nil {
-		return fmt.Errorf("already connected %v to peer %v", self.Id, id)
+		return
 	}
 	peer = self.setPeer(id, rrw)
 	close(peer.Connc)
 	defer close(peer.Readyc)
-	err := na.(ProtocolRunner).RunProtocol(self.Id, rrw, rw, peer)
-	if err != nil {
-		return fmt.Errorf("cannot run protocol (%v -> %v) %v", self.Id, id, err)
-	}
+	na.(*SimNode).RunProtocol(self, rrw, rw, peer)
 
 	// run protocol on remote node with self as peer
-	err = self.RunProtocol(id, rw, rrw, peer)
-	if err != nil {
-		return fmt.Errorf("cannot run protocol (%v -> %v): %v", id, self.Id, err)
-	}
+	self.RunProtocol(na.(*SimNode), rw, rrw, peer)
+}
+
+func (self *SimNode) PeerCount() int {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	return len(self.peers)
+}
+
+func (self *SimNode) NodeInfo() *p2p.NodeInfo {
+	return &p2p.NodeInfo{ID: self.Id.String()}
+}
+
+func (self *SimNode) PeersInfo() (info []*p2p.PeerInfo) {
 	return nil
 }
 
-func (self *SimNode) RunProtocol(id *NodeId, rw, rrw p2p.MsgReadWriter, peer *Peer) error {
-	if self.Run == nil {
+func (self *SimNode) RunProtocol(node *SimNode, rw, rrw p2p.MsgReadWriter, peer *Peer) {
+	id := node.Id
+	protocol := self.service.Protocols()[0]
```

Comment:
I was just trying to make a small change.

If there are multiple protocols running we need to have a mechanism to stop them if one of them exits before the others.

This is doable, but not necessary at the moment given we aren't testing anything which runs multiple protocols. We should just modify this when the time arises that we need to support multiple protocols.